File: schema/contract_schema.py
"""
Pydantic schema for validated contract data.

To add new fields or change validation rules, edit this file.
Future extension: create separate schemas for different document types
(e.g. ServiceContract, GrantAgreement) and select them in extract_data.py.
"""

from typing import Optional, List
from datetime import date
from pydantic import BaseModel


class ContractData(BaseModel):
    project_name: Optional[str] = None
    contract_partner: Optional[str] = None       # Drittmittelgeber / Auftraggeber
    organization: Optional[str] = None           # Lehrstuhl / Einrichtung
    address: Optional[str] = None
    contact_person: Optional[str] = None         # Projektleitung
    email: Optional[str] = None
    funding_amount: Optional[float] = None
    project_start_date: Optional[date] = None
    project_end_date: Optional[date] = None
    contract_reference: Optional[str] = None     # Förderkennzeichen / Vertragsnummer
    missing_fields: List[str] = []
    confidence_notes: Optional[str] = None

    def format_date(self, field_name: str) -> str:
        value = getattr(self, field_name)
        if value is None:
            return ""
        return value.strftime("%d.%m.%Y")

    def format_amount(self) -> str:
        if self.funding_amount is None:
            return ""
        # German number format: 1.234,56 €
        formatted = f"{self.funding_amount:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")
        return f"{formatted} €"

File: src/schema/contract_schema.ts
import { z } from 'zod'

export const ContractDataSchema = z.object({
  project_name: z.string().nullable().default(null),
  contract_partner: z.string().nullable().default(null),
  organization: z.string().nullable().default(null),
  address: z.string().nullable().default(null),
  contact_person: z.string().nullable().default(null),
  email: z.string().nullable().default(null),
  funding_amount: z.number().nullable().default(null),
  project_start_date: z.string().nullable().default(null), // ISO 8601: YYYY-MM-DD
  project_end_date: z.string().nullable().default(null),
  contract_reference: z.string().nullable().default(null),
  missing_fields: z.array(z.string()).default([]),
  confidence_notes: z.string().nullable().default(null),
})

export type ContractData = z.infer<typeof ContractDataSchema>

export function formatDate(date: string | null): string {
  if (!date) return ''
  // Split directly to avoid timezone-shift from new Date(isoString)
  const [year, month, day] = date.split('-').map(Number)
  return `${String(day).padStart(2, '0')}.${String(month).padStart(2, '0')}.${year}`
}

export function formatAmount(amount: number | null): string {
  if (amount === null) return ''
  return amount.toLocaleString('de-DE', { minimumFractionDigits: 2, maximumFractionDigits: 2 }) + ' €'
}
